File: examples/c_examples/05_slave_file.c
/**
 * FastDFS Slave File Example
 * 
 * This example demonstrates how to work with slave files in FastDFS.
 * Slave files are associated files linked to a master file, commonly used
 * for storing different versions or variants of the same content, such as:
 * - Image thumbnails (small, medium, large)
 * - Video transcodes (different resolutions/formats)
 * - Document previews
 * - Processed versions of original files
 * 
 * Copyright (C) 2024
 * License: GPL v3
 * 
 * USAGE:
 *   ./05_slave_file <config_file> <master_file> <slave_file> <prefix_name>
 * 
 * EXAMPLE:
 *   ./05_slave_file client.conf original.jpg thumbnail.jpg _150x150
 *   ./05_slave_file client.conf video.mp4 video_720p.mp4 _720p
 * 
 * EXPECTED OUTPUT:
 *   Master file uploaded!
 *   File ID: group1/M00/00/00/wKgBcGXxxx.jpg
 *   
 *   Slave file uploaded!
 *   Slave File ID: group1/M00/00/00/wKgBcGXxxx_150x150.jpg
 *   
 *   Slave file downloaded successfully!
 *   Downloaded to: downloaded_slave_150x150.jpg
 * 
 * COMMON PITFALLS:
 *   1. Prefix naming - Must start with underscore or hyphen (e.g., _thumb, -small)
 *   2. Master file must exist - Upload master before slave
 *   3. Same storage server - Slave must be on same server as master
 *   4. Deleting master - Deleting master doesn't auto-delete slaves
 *   5. Prefix uniqueness - Different slaves need different prefixes
 *   6. File extension - Slave can have different extension than master
 * 
 * KEY CONCEPTS:
 *   - Slave files are stored on the same storage server as the master
 *   - Slave filename = master_filename + prefix + extension
 *   - Multiple slaves can be attached to one master
 *   - Slaves are independent files but logically linked
 *   - Useful for multi-resolution images, video transcodes, etc.
 *   - Slaves don't automatically inherit master's metadata
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/stat.h>
#include "fdfs_client.h"
#include "fdfs_global.h"
#include "tracker_types.h"
#include "storage_client.h"
#include "fastcommon/logger.h"

/**
 * Print usage information
 */
void print_usage(const char *program_name)
{
    printf("FastDFS Slave File Example\n\n");
    printf("Usage: %s <config_file> <master_file> <slave_file> <prefix_name>\n\n", 
           program_name);
    printf("Arguments:\n");
    printf("  config_file  Path to FastDFS client configuration file\n");
    printf("  master_file  Path to the master file to upload\n");
    printf("  slave_file   Path to the slave file to upload\n");
    printf("  prefix_name  Prefix for slave file (e.g., _150x150, _thumb, -small)\n\n");
    printf("Example:\n");
    printf("  %s client.conf photo.jpg thumbnail.jpg _150x150\n", program_name);
    printf("  %s client.conf video.mp4 video_hd.mp4 _720p\n\n", program_name);
    printf("What this does:\n");
    printf("  1. Uploads master file to FastDFS\n");
    printf("  2. Uploads slave file linked to master with prefix\n");
    printf("  3. Downloads the slave file to verify\n");
    printf("  4. Demonstrates querying slave file info\n");
}

/**
 * Validate that the file exists and is readable
 */
int validate_file(const char *filepath)
{
    struct stat stat_buf;
    
    if (stat(filepath, &stat_buf) != 0) {
        fprintf(stderr, "ERROR: Cannot access file '%s': %s\n", 
                filepath, strerror(errno));
        return errno;
    }
    
    if (!S_ISREG(stat_buf.st_mode)) {
        fprintf(stderr, "ERROR: '%s' is not a regular file\n", filepath);
        return EINVAL;
    }
    
    return 0;
}

/**
 * Validate slave file prefix
 * Prefix should start with underscore or hyphen
 */
int validate_prefix(const char *prefix)
{
    if (prefix == NULL || strlen(prefix) == 0) {
        fprintf(stderr, "ERROR: Prefix cannot be empty\n");
        return EINVAL;
    }
    
    if (prefix[0] != '_' && prefix[0] != '-') {
        fprintf(stderr, "WARNING: Prefix '%s' doesn't start with _ or -\n", prefix);
        fprintf(stderr, "         This is recommended but not required\n");
    }
    
    if (strlen(prefix) > 64) {
        fprintf(stderr, "ERROR: Prefix too long (max 64 characters)\n");
        return EINVAL;
    }
    
    return 0;
}

int main(int argc, char *argv[])
{
    char *conf_filename;
    char *master_filename;
    char *slave_filename;
    char *prefix_name;
    ConnectionInfo *pTrackerServer;
    ConnectionInfo *pStorageServer;
    ConnectionInfo storageServer;
    int result;
    char group_name[FDFS_GROUP_NAME_MAX_LEN + 1];
    char master_remote_filename[256];
    char slave_remote_filename[256];
    char master_file_id[128];
    char slave_file_id[128];
    const char *master_ext;
    const char *slave_ext;
    int store_path_index;
    FDFSFileInfo file_info;
    char download_filename[256];
    int64_t file_size;
    
    /* ========================================
     * STEP 1: Parse and validate arguments
     * ======================================== */
    if (argc != 5) {
        print_usage(argv[0]);
        return 1;
    }
    
    conf_filename = argv[1];
    master_filename = argv[2];
    slave_filename = argv[3];
    prefix_name = argv[4];
    
    /* Validate input files */
    if ((result = validate_file(master_filename)) != 0) {
        return result;
    }
    if ((result = validate_file(slave_filename)) != 0) {
        return result;
    }
    if ((result = validate_prefix(prefix_name)) != 0) {
        return result;
    }
    
    printf("=== FastDFS Slave File Example ===\n");
    printf("Config file: %s\n", conf_filename);
    printf("Master file: %s\n", master_filename);
    printf("Slave file: %s\n", slave_filename);
    printf("Prefix name: %s\n\n", prefix_name);
    
    /* ========================================
     * STEP 2: Initialize FastDFS client
     * ======================================== */
    log_init();
    
    printf("Initializing FastDFS client...\n");
    if ((result = fdfs_client_init(conf_filename)) != 0) {
        fprintf(stderr, "ERROR: Failed to initialize FastDFS client\n");
        fprintf(stderr, "Error code: %d, Error info: %s\n", 
                result, STRERROR(result));
        return result;
    }
    printf("✓ Client initialized successfully\n\n");
    
    /* ========================================
     * STEP 3: Connect to tracker server
     * ======================================== */
    printf("Connecting to tracker server...\n");
    pTrackerServer = tracker_get_connection();
    if (pTrackerServer == NULL) {
        result = errno != 0 ? errno : ECONNREFUSED;
        fprintf(stderr, "ERROR: Failed to connect to tracker server\n");
        fprintf(stderr, "Error code: %d, Error info: %s\n", 
                result, STRERROR(result));
        fdfs_client_destroy();
        return result;
    }
    printf("✓ Connected to tracker server: %s:%d\n\n", 
           pTrackerServer->ip_addr, pTrackerServer->port);
    
    /* ========================================
     * STEP 4: Query storage server
     * ======================================== */
    printf("Querying storage server for upload...\n");
    store_path_index = 0;
    memset(group_name, 0, sizeof(group_name));
    
    result = tracker_query_storage_store(pTrackerServer, 
                                         &storageServer, 
                                         group_name, 
                                         &store_path_index);
    if (result != 0) {
        fprintf(stderr, "ERROR: Failed to query storage server\n");
        fprintf(stderr, "Error code: %d, Error info: %s\n", 
                result, STRERROR(result));
        tracker_close_connection_ex(pTrackerServer, true);
        fdfs_client_destroy();
        return result;
    }
    
    printf("✓ Storage server assigned: %s:%d (group: %s)\n\n", 
           storageServer.ip_addr, storageServer.port, group_name);
    
    /* ========================================
     * STEP 5: Connect to storage server
     * ======================================== */
    printf("Connecting to storage server...\n");
    pStorageServer = tracker_make_connection(&storageServer, &result);
    if (pStorageServer == NULL) {
        fprintf(stderr, "ERROR: Failed to connect to storage server\n");
        fprintf(stderr, "Error code: %d, Error info: %s\n", 
                result, STRERROR(result));
        tracker_close_connection_ex(pTrackerServer, true);
        fdfs_client_destroy();
        return result;
    }
    printf("✓ Connected to storage server\n\n");
    
    /* ========================================
     * STEP 6: Upload MASTER file
     * ======================================== */
    printf("=== PHASE 1: Upload Master File ===\n");
    printf("Uploading master file '%s'...\n", master_filename);
    
    master_ext = fdfs_get_file_ext_name(master_filename);
    
    /* Upload master file using standard upload function */
    result = storage_upload_by_filename(pTrackerServer,
                                        pStorageServer,
                                        store_path_index,
                                        master_filename,
                                        master_ext,
                                        NULL,  /* No metadata */
                                        0,     /* Metadata count */
                                        group_name,
                                        master_remote_filename);
    
    if (result != 0) {
        fprintf(stderr, "ERROR: Failed to upload master file\n");
        fprintf(stderr, "Error code: %d, Error info: %s\n", 
                result, STRERROR(result));
        goto cleanup;
    }
    
    /* Construct master file ID */
    snprintf(master_file_id, sizeof(master_file_id), 
             "%s/%s", group_name, master_remote_filename);
    
    printf("✓ Master file uploaded successfully!\n");
    printf("  Master File ID: %s\n", master_file_id);
    
    /* Get master file info */
    result = fdfs_get_file_info(group_name, master_remote_filename, &file_info);
    if (result == 0) {
        printf("  Master file size: %lld bytes\n", (long long)file_info.file_size);
        printf("  CRC32: %u\n", file_info.crc32);
    }
    printf("\n");
    
    /* ========================================
     * STEP 7: Upload SLAVE file
     * ======================================== */
    printf("=== PHASE 2: Upload Slave File ===\n");
    printf("Uploading slave file '%s' with prefix '%s'...\n", 
           slave_filename, prefix_name);
    
    slave_ext = fdfs_get_file_ext_name(slave_filename);
    
    /* Upload slave file linked to master
     * IMPORTANT: Use storage_upload_slave_by_filename() to create a slave file
     * 
     * The slave file will be stored on the same storage server as the master
     * and its filename will be: master_filename + prefix + extension
     * 
     * Parameters:
     *   - pTrackerServer: tracker connection
     *   - pStorageServer: storage connection (can be NULL)
     *   - slave_filename: local path to slave file
     *   - master_remote_filename: the master file's remote filename
     *   - prefix_name: prefix to identify this slave (e.g., _150x150)
     *   - slave_ext: file extension for slave file
     *   - NULL, 0: metadata (optional)
     *   - group_name: group name (same as master)
     *   - slave_remote_filename: output - generated slave filename
     */
    result = storage_upload_slave_by_filename(pTrackerServer,
                                              pStorageServer,
                                              slave_filename,
                                              master_remote_filename,
                                              prefix_name,
                                              slave_ext,
                                              NULL,  /* No metadata */
                                              0,     /* Metadata count */
                                              group_name,
                                              slave_remote_filename);
    
    if (result != 0) {
        fprintf(stderr, "ERROR: Failed to upload slave file\n");
        fprintf(stderr, "Error code: %d, Error info: %s\n", 
                result, STRERROR(result));
        fprintf(stderr, "\nPossible causes:\n");
        fprintf(stderr, "  - Master file doesn't exist\n");
        fprintf(stderr, "  - Invalid prefix format\n");
        fprintf(stderr, "  - Storage server connection lost\n");
        fprintf(stderr, "  - Insufficient disk space\n");
        goto cleanup;
    }
    
    /* Construct slave file ID */
    snprintf(slave_file_id, sizeof(slave_file_id), 
             "%s/%s", group_name, slave_remote_filename);
    
    printf("✓ Slave file uploaded successfully!\n");
    printf("  Slave File ID: %s\n", slave_file_id);
    
    /* Get slave file info */
    result = fdfs_get_file_info(group_name, slave_remote_filename, &file_info);
    if (result == 0) {
        printf("  Slave file size: %lld bytes\n", (long long)file_info.file_size);
        printf("  CRC32: %u\n", file_info.crc32);
        printf("  Source IP: %s\n", file_info.source_ip_addr);
    }
    printf("\n");
    
    /* ========================================
     * STEP 8: Demonstrate filename relationship
     * ======================================== */
    printf("=== PHASE 3: Filename Relationship ===\n");
    printf("Master filename: %s\n", master_remote_filename);
    printf("Slave filename:  %s\n", slave_remote_filename);
    printf("\nNotice how slave filename is constructed:\n");
    printf("  Base: %s\n", master_remote_filename);
    printf("  + Prefix: %s\n", prefix_name);
    printf("  + Extension: .%s\n", slave_ext ? slave_ext : "(none)");
    printf("\n");
    
    /* ========================================
     * STEP 9: Download slave file to verify
     * ======================================== */
    printf("=== PHASE 4: Download Slave File ===\n");
    
    /* Construct download filename */
    snprintf(download_filename, sizeof(download_filename), 
             "downloaded_slave%s.%s", prefix_name, slave_ext ? slave_ext : "dat");
    
    printf("Downloading slave file to '%s'...\n", download_filename);
    
    /* Download the slave file
     * Use the same download function as for regular files
     */
    result = storage_download_file_to_file(pTrackerServer,
                                           pStorageServer,
                                           group_name,
                                           slave_remote_filename,
                                           download_filename,
                                           &file_size);
    
    if (result != 0) {
        fprintf(stderr, "WARNING: Failed to download slave file\n");
        fprintf(stderr, "Error code: %d, Error info: %s\n", 
                result, STRERROR(result));
        /* Non-fatal, continue */
    } else {
        printf("✓ Slave file downloaded successfully!\n");
        printf("  Downloaded to: %s\n", download_filename);
        printf("  Downloaded size: %lld bytes\n", (long long)file_size);
    }
    printf("\n");
    
    /* ========================================
     * STEP 10: Display summary and use cases
     * ======================================== */
    printf("=== Summary ===\n");
    printf("Slave file operations completed successfully!\n\n");
    
    printf("Files created:\n");
    printf("  Master: %s\n", master_file_id);
    printf("  Slave:  %s\n", slave_file_id);
    
    printf("\nCommon use cases for slave files:\n");
    printf("  1. Image thumbnails:\n");
    printf("     - master: original.jpg\n");
    printf("     - slave: original_150x150.jpg (thumbnail)\n");
    printf("     - slave: original_800x600.jpg (medium)\n");
    printf("     - slave: original_1920x1080.jpg (large)\n\n");
    
    printf("  2. Video transcoding:\n");
    printf("     - master: video.mp4 (original 4K)\n");
    printf("     - slave: video_1080p.mp4\n");
    printf("     - slave: video_720p.mp4\n");
    printf("     - slave: video_480p.mp4\n\n");
    
    printf("  3. Document formats:\n");
    printf("     - master: document.pdf\n");
    printf("     - slave: document_preview.jpg\n");
    printf("     - slave: document_text.txt\n\n");
    
    printf("  4. Audio quality variants:\n");
    printf("     - master: song.flac (lossless)\n");
    printf("     - slave: song_320k.mp3\n");
    printf("     - slave: song_128k.mp3\n\n");
    
    printf("Best practices:\n");
    printf("  ✓ Use descriptive prefixes (e.g., _thumb, _720p, _preview)\n");
    printf("  ✓ Upload master first, then slaves\n");
    printf("  ✓ Keep prefix naming consistent across your application\n");
    printf("  ✓ Document your prefix conventions\n");
    printf("  ✓ Consider slave file lifecycle with master deletion\n");
    
    result = 0;

cleanup:
    /* ========================================
     * STEP 11: Cleanup
     * ======================================== */
    if (pStorageServer != NULL) {
        tracker_close_connection_ex(pStorageServer, result != 0);
    }
    if (pTrackerServer != NULL) {
        tracker_close_connection_ex(pTrackerServer, result != 0);
    }
    fdfs_client_destroy();
    
    printf("\n=== Cleanup Complete ===\n");
    
    return result;
}

File: examples/php_examples/05_slave_file.php
<?php
/**
 * FastDFS Slave File Operations Example
 * 
 * This example demonstrates how to work with slave files in FastDFS.
 * Slave files are derived files (like thumbnails, compressed versions, or 
 * different formats) that are associated with a master file.
 * 
 * Key concepts:
 * - Master file: The original uploaded file
 * - Slave file: A derived/related file linked to the master
 * - Prefix/Suffix: Used to identify the relationship between files
 * 
 * Common use cases:
 * - Image thumbnails (small, medium, large versions)
 * - Video transcoding (different resolutions/formats)
 * - Document conversions (PDF to images, etc.)
 * - Compressed versions of files
 * 
 * Prerequisites:
 * - FastDFS PHP extension installed (php-fastdfs)
 * - FastDFS tracker server running and accessible
 * - Master file already uploaded to FastDFS
 */

// FastDFS tracker server configuration
define('TRACKER_HOST', '127.0.0.1');
define('TRACKER_PORT', 22122);

/**
 * Initialize connection to FastDFS tracker server
 * 
 * @return resource|false FastDFS tracker connection or false on failure
 */
function initializeFastDFS() {
    try {
        $tracker = fastdfs_tracker_make_connection(TRACKER_HOST, TRACKER_PORT);
        
        if (!$tracker) {
            throw new Exception("Failed to connect to tracker server");
        }
        
        echo "✓ Connected to FastDFS tracker server\n";
        return $tracker;
        
    } catch (Exception $e) {
        echo "✗ Connection error: " . $e->getMessage() . "\n";
        return false;
    }
}

/**
 * Upload a master file to FastDFS
 * 
 * This is the primary/original file that slave files will be associated with.
 * 
 * @param resource $tracker FastDFS tracker connection
 * @param string $localFilePath Path to the local file
 * @param array $metadata Optional metadata
 * @return string|false Master file ID on success, false on failure
 */
function uploadMasterFile($tracker, $localFilePath, $metadata = []) {
    try {
        echo "\n--- Uploading Master File ---\n";
        echo "File: $localFilePath\n";
        
        if (!file_exists($localFilePath)) {
            throw new Exception("File not found: $localFilePath");
        }
        
        $fileSize = filesize($localFilePath);
        $fileExtension = pathinfo($localFilePath, PATHINFO_EXTENSION);
        
        echo "Size: $fileSize bytes\n";
        echo "Extension: $fileExtension\n";
        
        // Upload the master file
        $fileId = fastdfs_storage_upload_by_filename(
            $localFilePath,
            $fileExtension,
            $metadata,
            [],
            TRACKER_HOST,
            TRACKER_PORT
        );
        
        if (!$fileId) {
            throw new Exception("Master file upload failed");
        }
        
        echo "✓ Master file uploaded successfully!\n";
        echo "Master File ID: $fileId\n";
        
        return $fileId;
        
    } catch (Exception $e) {
        echo "✗ Upload error: " . $e->getMessage() . "\n";
        return false;
    }
}

/**
 * Upload a slave file associated with a master file
 * 
 * Slave files are linked to master files using a prefix/suffix naming convention.
 * The slave file is stored in the same group as the master file.
 * 
 * @param resource $tracker FastDFS tracker connection
 * @param string $masterFileId Master file ID
 * @param string $localFilePath Path to the slave file
 * @param string $prefixName Prefix to identify the slave file (e.g., "_thumb", "_small", "_150x150")
 * @param string $fileExtension File extension for slave file
 * @return string|false Slave file ID on success, false on failure
 */
function uploadSlaveFile($tracker, $masterFileId, $localFilePath, $prefixName, $fileExtension = '') {
    try {
        echo "\n--- Uploading Slave File ---\n";
        echo "Master File ID: $masterFileId\n";
        echo "Slave File: $localFilePath\n";
        echo "Prefix: $prefixName\n";
        
        if (!file_exists($localFilePath)) {
            throw new Exception("Slave file not found: $localFilePath");
        }
        
        if (empty($fileExtension)) {
            $fileExtension = pathinfo($localFilePath, PATHINFO_EXTENSION);
        }
        
        $fileSize = filesize($localFilePath);
        echo "Size: $fileSize bytes\n";
        echo "Extension: $fileExtension\n";
        
        // Upload slave file linked to master file
        // The slave file will be stored with a reference to the master file
        $slaveFileId = fastdfs_storage_upload_slave_by_filename(
            $localFilePath,
            $masterFileId,
            $prefixName,
            $fileExtension,
            [],  // Metadata
            [],  // Group name
            TRACKER_HOST,
            TRACKER_PORT
        );
        
        if (!$slaveFileId) {
            throw new Exception("Slave file upload failed");
        }
        
        echo "✓ Slave file uploaded successfully!\n";
        echo "Slave File ID: $slaveFileId\n";
        
        return $slaveFileId;
        
    } catch (Exception $e) {
        echo "✗ Slave upload error: " . $e->getMessage() . "\n";
        return false;
    }
}

/**
 * Upload a slave file from memory buffer
 * 
 * @param resource $tracker FastDFS tracker connection
 * @param string $masterFileId Master file ID
 * @param string $content Slave file content
 * @param string $prefixName Prefix to identify the slave file
 * @param string $fileExtension File extension
 * @return string|false Slave file ID on success, false on failure
 */
function uploadSlaveFromBuffer($tracker, $masterFileId, $content, $prefixName, $fileExtension) {
    try {
        echo "\n--- Uploading Slave File from Buffer ---\n";
        echo "Master File ID: $masterFileId\n";
        echo "Content size: " . strlen($content) . " bytes\n";
        echo "Prefix: $prefixName\n";
        echo "Extension: $fileExtension\n";
        
        // Upload slave file content from memory
        $slaveFileId = fastdfs_storage_upload_slave_by_filebuff(
            $content,
            $masterFileId,
            $prefixName,
            $fileExtension,
            [],  // Metadata
            [],  // Group name
            TRACKER_HOST,
            TRACKER_PORT
        );
        
        if (!$slaveFileId) {
            throw new Exception("Slave buffer upload failed");
        }
        
        echo "✓ Slave file uploaded from buffer!\n";
        echo "Slave File ID: $slaveFileId\n";
        
        return $slaveFileId;
        
    } catch (Exception $e) {
        echo "✗ Slave buffer upload error: " . $e->getMessage() . "\n";
        return false;
    }
}

/**
 * Download a slave file to local disk
 * 
 * @param resource $tracker FastDFS tracker connection
 * @param string $slaveFileId Slave file ID
 * @param string $localPath Local path to save the file
 * @return bool True on success, false on failure
 */
function downloadSlaveFile($tracker, $slaveFileId, $localPath) {
    try {
        echo "\n--- Downloading Slave File ---\n";
        echo "Slave File ID: $slaveFileId\n";
        echo "Save to: $localPath\n";
        
        // Download slave file (same as regular file download)
        $result = fastdfs_storage_download_file_to_file(
            $slaveFileId,
            $localPath,
            TRACKER_HOST,
            TRACKER_PORT
        );
        
        if (!$result) {
            throw new Exception("Download failed");
        }
        
        if (!file_exists($localPath)) {
            throw new Exception("File was not saved to disk");
        }
        
        $fileSize = filesize($localPath);
        echo "✓ Download successful!\n";
        echo "File size: $fileSize bytes\n";
        
        return true;
        
    } catch (Exception $e) {
        echo "✗ Download error: " . $e->getMessage() . "\n";
        return false;
    }
}

/**
 * Download slave file to memory buffer
 * 
 * @param resource $tracker FastDFS tracker connection
 * @param string $slaveFileId Slave file ID
 * @return string|false File content on success, false on failure
 */
function downloadSlaveToBuffer($tracker, $slaveFileId) {
    try {
        echo "\n--- Downloading Slave File to Buffer ---\n";
        echo "Slave File ID: $slaveFileId\n";
        
        $content = fastdfs_storage_download_file_to_buff(
            $slaveFileId,
            TRACKER_HOST,
            TRACKER_PORT
        );
        
        if ($content === false) {
            throw new Exception("Download to buffer failed");
        }
        
        echo "✓ Download successful!\n";
        echo "Content size: " . strlen($content) . " bytes\n";
        
        return $content;
        
    } catch (Exception $e) {
        echo "✗ Download error: " . $e->getMessage() . "\n";
        return false;
    }
}

/**
 * Delete a slave file
 * 
 * Note: Deleting a master file does NOT automatically delete its slave files.
 * You must delete slave files explicitly.
 * 
 * @param resource $tracker FastDFS tracker connection
 * @param string $slaveFileId Slave file ID to delete
 * @return bool True on success, false on failure
 */
function deleteSlaveFile($tracker, $slaveFileId) {
    try {
        echo "\n--- Deleting Slave File ---\n";
        echo "Slave File ID: $slaveFileId\n";
        
        // Delete the slave file
        $result = fastdfs_storage_delete_file(
            $slaveFileId,
            TRACKER_HOST,
            TRACKER_PORT
        );
        
        if (!$result) {
            throw new Exception("Delete operation failed");
        }
        
        echo "✓ Slave file deleted successfully!\n";
        return true;
        
    } catch (Exception $e) {
        echo "✗ Delete error: " . $e->getMessage() . "\n";
        return false;
    }
}

/**
 * Create a test image file for demonstration
 * 
 * @param string $filename Filename to create
 * @param int $width Image width
 * @param int $height Image height
 * @param string $text Text to display on image
 * @return string Path to created file
 */
function createTestImage($filename, $width, $height, $text) {
    $filepath = __DIR__ . '/' . $filename;
    
    // Create a simple image using GD library if available
    if (function_exists('imagecreate')) {
        $image = imagecreate($width, $height);
        $bgColor = imagecolorallocate($image, 200, 200, 200);
        $textColor = imagecolorallocate($image, 0, 0, 0);
        
        imagestring($image, 5, 10, $height / 2 - 10, $text, $textColor);
        imagejpeg($image, $filepath);
        imagedestroy($image);
    } else {
        // Fallback: create a text file if GD is not available
        file_put_contents($filepath, "Image placeholder: $text\nSize: {$width}x{$height}");
    }
    
    return $filepath;
}

/**
 * Demonstrate image thumbnail scenario
 * 
 * This shows a common use case: uploading an original image and multiple
 * thumbnail versions as slave files.
 * 
 * @param resource $tracker FastDFS tracker connection
 */
function demonstrateImageThumbnails($tracker) {
    echo "\n" . str_repeat("=", 70) . "\n";
    echo "=== Image Thumbnail Scenario ===\n";
    echo str_repeat("=", 70) . "\n";
    
    // Create original image
    $originalImage = createTestImage('original.jpg', 800, 600, 'Original Image');
    echo "\nCreated test image: $originalImage\n";
    
    // Upload master file (original image)
    $metadata = [
        'type' => 'image',
        'original_name' => 'photo.jpg',
        'upload_date' => date('Y-m-d H:i:s')
    ];
    
    $masterFileId = uploadMasterFile($tracker, $originalImage, $metadata);
    
    if (!$masterFileId) {
        return;
    }
    
    // Create and upload thumbnail versions as slave files
    $thumbnails = [
        ['prefix' => '_thumb_small', 'width' => 150, 'height' => 150, 'label' => 'Small Thumb'],
        ['prefix' => '_thumb_medium', 'width' => 300, 'height' => 300, 'label' => 'Medium Thumb'],
        ['prefix' => '_thumb_large', 'width' => 600, 'height' => 600, 'label' => 'Large Thumb']
    ];
    
    $slaveFileIds = [];
    
    foreach ($thumbnails as $thumb) {
        // Create thumbnail image
        $thumbFile = createTestImage(
            "thumb_{$thumb['width']}x{$thumb['height']}.jpg",
            $thumb['width'],
            $thumb['height'],
            $thumb['label']
        );
        
        // Upload as slave file
        $slaveId = uploadSlaveFile(
            $tracker,
            $masterFileId,
            $thumbFile,
            $thumb['prefix'],
            'jpg'
        );
        
        if ($slaveId) {
            $slaveFileIds[$thumb['prefix']] = $slaveId;
        }
        
        // Clean up local thumbnail file
        @unlink($thumbFile);
    }
    
    // Summary
    echo "\n--- Image Upload Summary ---\n";
    echo "Master File ID: $masterFileId\n";
    echo "\nSlave Files (Thumbnails):\n";
    foreach ($slaveFileIds as $prefix => $fileId) {
        echo "  $prefix: $fileId\n";
    }
    
    // Clean up original image
    @unlink($originalImage);
    
    return [
        'master' => $masterFileId,
        'slaves' => $slaveFileIds
    ];
}

/**
 * Demonstrate document conversion scenario
 * 
 * Shows uploading a document and its converted versions (e.g., PDF and images)
 * 
 * @param resource $tracker FastDFS tracker connection
 */
function demonstrateDocumentConversion($tracker) {
    echo "\n" . str_repeat("=", 70) . "\n";
    echo "=== Document Conversion Scenario ===\n";
    echo str_repeat("=", 70) . "\n";
    
    // Create master document
    $docContent = "FastDFS Slave File Example\n\n";
    $docContent .= "This is a sample document that demonstrates slave file operations.\n";
    $docContent .= "In a real scenario, this could be a Word document, PDF, or other format.\n";
    $docContent .= "\nCreated: " . date('Y-m-d H:i:s') . "\n";
    
    $docFile = __DIR__ . '/document.txt';
    file_put_contents($docFile, $docContent);
    
    // Upload master document
    $masterFileId = uploadMasterFile($tracker, $docFile, [
        'type' => 'document',
        'format' => 'text'
    ]);
    
    if (!$masterFileId) {
        @unlink($docFile);
        return;
    }
    
    // Create and upload converted versions as slave files
    
    // 1. Compressed version
    $compressedContent = gzcompress($docContent);
    $compressedId = uploadSlaveFromBuffer(
        $tracker,
        $masterFileId,
        $compressedContent,
        '_compressed',
        'gz'
    );
    
    // 2. HTML version
    $htmlContent = "<html><body><pre>" . htmlspecialchars($docContent) . "</pre></body></html>";
    $htmlId = uploadSlaveFromBuffer(
        $tracker,
        $masterFileId,
        $htmlContent,
        '_html',
        'html'
    );
    
    // 3. JSON metadata version
    $jsonContent = json_encode([
        'title' => 'FastDFS Example Document',
        'content' => $docContent,
        'created' => date('Y-m-d H:i:s'),
        'length' => strlen($docContent)
    ], JSON_PRETTY_PRINT);
    
    $jsonId = uploadSlaveFromBuffer(
        $tracker,
        $masterFileId,
        $jsonContent,
        '_metadata',
        'json'
    );
    
    // Summary
    echo "\n--- Document Conversion Summary ---\n";
    echo "Master Document ID: $masterFileId\n";
    echo "\nConverted Versions (Slave Files):\n";
    if ($compressedId) echo "  Compressed (_compressed.gz): $compressedId\n";
    if ($htmlId) echo "  HTML (_html.html): $htmlId\n";
    if ($jsonId) echo "  JSON Metadata (_metadata.json): $jsonId\n";
    
    // Test downloading a slave file
    if ($htmlId) {
        $content = downloadSlaveToBuffer($tracker, $htmlId);
        if ($content) {
            echo "\n--- Downloaded HTML Version (first 200 chars) ---\n";
            echo substr($content, 0, 200) . "...\n";
        }
    }
    
    // Clean up
    @unlink($docFile);
    
    return [
        'master' => $masterFileId,
        'slaves' => [
            'compressed' => $compressedId,
            'html' => $htmlId,
            'json' => $jsonId
        ]
    ];
}

/**
 * Main execution function
 */
function main() {
    echo "=== FastDFS Slave File Operations Example ===\n\n";
    
    echo "Slave files allow you to store related/derived files linked to a master file.\n";
    echo "Common use cases:\n";
    echo "  - Image thumbnails of different sizes\n";
    echo "  - Video files in different resolutions\n";
    echo "  - Document format conversions\n";
    echo "  - Compressed versions of files\n\n";
    
    // Initialize connection
    $tracker = initializeFastDFS();
    if (!$tracker) {
        exit(1);
    }
    
    // Scenario 1: Image thumbnails
    $imageResult = demonstrateImageThumbnails($tracker);
    
    // Scenario 2: Document conversion
    $docResult = demonstrateDocumentConversion($tracker);
    
    // Final summary
    echo "\n" . str_repeat("=", 70) . "\n";
    echo "=== Summary ===\n";
    echo str_repeat("=", 70) . "\n";
    
    echo "\nOperations demonstrated:\n";
    echo "✓ Upload master file\n";
    echo "✓ Upload slave files from disk\n";
    echo "✓ Upload slave files from buffer\n";
    echo "✓ Download slave files\n";
    echo "✓ Multiple slave files per master\n";
    
    echo "\nKey Points:\n";
    echo "- Slave files are linked to master files using prefix/suffix\n";
    echo "- Multiple slave files can be associated with one master file\n";
    echo "- Slave files are stored in the same group as the master file\n";
    echo "- Deleting master file does NOT auto-delete slave files\n";
    echo "- Slave files can have different extensions than master file\n";
    echo "- Use descriptive prefixes to identify slave file purpose\n";
    
    echo "\nFile IDs for testing:\n";
    if ($imageResult) {
        echo "\nImage Example:\n";
        echo "  Master: {$imageResult['master']}\n";
        foreach ($imageResult['slaves'] as $prefix => $id) {
            echo "  Slave $prefix: $id\n";
        }
    }
    
    if ($docResult) {
        echo "\nDocument Example:\n";
        echo "  Master: {$docResult['master']}\n";
        foreach ($docResult['slaves'] as $type => $id) {
            echo "  Slave $type: $id\n";
        }
    }
    
    // Close tracker connection
    fastdfs_tracker_close_connection($tracker);
    echo "\n✓ Connection closed\n";
}

// Run the example
main();

?>
